air/src/aux_table/mod.rs

```diff
@@ -0,0 +1,198 @@
+/// bitwise trace ends.
+pub const S2_COL_IDX: usize = AUX_TRACE_OFFSET + 2;
+/// The first column of the memory co-processor.
+pub const MEMORY_TRACE_OFFSET: usize = S2_COL_IDX + 1;
+/// The degree of the virtual flag that selects the memory section of the auxiliary table.
+pub const MEMORY_FLAG_DEGREE: usize = 3;
+
+// AUXILIARY TABLE TRANSITION CONSTRAINTS
+// ================================================================================================
+struct TransitionConstraintRange {
+    memory: Range<usize>,
+}
+
+impl TransitionConstraintRange {
+    fn new(start: usize, memory: usize) -> Self {
+        Self {
+            memory: create_range(start, memory),
+        }
+    }
+
+    fn memory_start(&self) -> usize {
+        self.memory.start
+    }
+}
+
+pub struct AuxTableConstraints<E: FieldElement> {
+    transitions: Vec<TransitionConstraint<E>>,
+    constraint_ranges: TransitionConstraintRange,
+    memory: MemoryConstraints<E>,
+}
+
+impl<E: FieldElement> AuxTableConstraints<E> {
+    pub fn new() -> Self {
+        // Initialize the co-processors.
+        let memory = MemoryConstraints::new();
+
+        // Define the transition constraints.
+        let transitions = build_transition_constraints();
+
+        // Define the co-processor constraint ranges.
+        let constraint_ranges = TransitionConstraintRange::new(
+            transitions.len(),
+            memory.get_transition_constraint_count(),
+        );
+
+        AuxTableConstraints {
+            transitions,
+            constraint_ranges,
+            memory,
+        }
+    }
+
+    fn memory_start(&self) -> usize {
+        self.constraint_ranges.memory_start()
+    }
+}
+
+impl<E: FieldElement> ProcessorTransitions<E> for AuxTableConstraints<E> {
+    fn transitions(&self) -> &Vec<TransitionConstraint<E>> {
+        &self.transitions
+    }
+
+    fn get_transition_constraint_count(&self) -> usize {
+        self.transitions().len() + self.memory.get_transition_constraint_count()
+    }
+
+    fn get_transition_constraint_degrees(&self) -> Vec<TransitionConstraintDegree> {
+        let mut aux_table_degrees: Vec<TransitionConstraintDegree> = self
+            .transitions()
+            .iter()
+            .map(|constraint| TransitionConstraintDegree::new(constraint.degree()))
+            .collect();
+        aux_table_degrees.append(&mut self.memory.get_transition_constraint_degrees());
+
+        aux_table_degrees
+    }
+
+    fn enforce_constraints(&self, frame: &EvaluationFrame<E>, result: &mut [E]) {
+        // --- auxiliary table management ---------------------------------------------------------
+        for (index, constraint) in self.transitions().iter().enumerate() {
+            result[index] = constraint.evaluate(frame);
+        }
+
+        // --- memory -----------------------------------------------------------------------------
+        self.memory
+            .enforce_constraints(frame, &mut result[self.memory_start()..]);
```

Comment:
I took this approach in a previous commit as well. However, I removed that when I switched to the approach with the processor constraints, because most of the structs using this trait won't need it, so passing it here didn't match the signature. If you don't like the current handling accessing the AuxTable frame extension then I could pass the flag when instantiating the memory constraints instead. What do you think?